content/blog/prototyping/index.md

```diff
@@ -0,0 +1,732 @@
+    note.content.push_str(" with additions");
+});
+```
+
+([Playground](https://play.rust-lang.org/?version=stable&mode=debug&edition=2021&gist=4b93a53ebc1d7ee6bc2b39c91543fba7))
+
+### Use A Flat Module Hierarchy
+
+Your `main.rs` is a great place for the entire prototype. 
+That's fine because you can always split it up later as soon as this becomes "uncomfortable."
+By then, you have a clear vision of what things should be called and how they should interact.
+
+I keep modules in the same file until they grow too large:
+
+```rust
+mod foo {
+    pub fn bar() {
+        println!("Hello from foo::bar");
+    }
+}
+
+mod baz {
+    pub fn qux() {
+        println!("Hello from baz::qux");
+    }
+}
+
+fn main() {
+    foo::bar();
+    baz::qux();
+}
+```
+
+This allows me to quickly move around in the codebase (as it's just one file) and also move code around by just cutting and pasting from one scope to another.
+It's very cheap.
+The individual modules can be moved to their own files later.
+
+On top of that, the individual modules can be treated like a black box.
+When I'm working on a library, this helps me to quickly test the API and see if the naming makes sense and if the modules are structured correctly.
+E.g. I might have a `config` module with a `load` function that returns a `Config` struct, which gets used by the `server` module to start the server. 
+
+```rust
+mod config {
+    pub struct Config {
+        pub port: u16,
+    }
+
+    pub fn load() -> Config {
+        Config { port: 8080 }
+    }
+}
+
+mod server {
+    use crate::config;
+
+    pub struct Server {
+        config: config::Config,
+    }
+
+    impl Server {
+        pub fn new(config: config::Config) -> Self {
+            Server { config }
+        }
+
+        pub fn start(&self) {
+            println!("Starting server on port {}", self.config.port);
+        }
+    }
+}
+
+use server::Server;
+
+fn main() {
+    let config = config::load();
+    let server = Server::new(config);
+    server.start();
+}
+```
+
+In general, I try to keep a flat hierarchy in the beginning and only introduce more structure when it becomes necessary.
+See also [Matklad's article on large Rust workspaces](https://matklad.github.io/2021/08/22/large-rust-workspaces.html).
+
+## Summary
+
+That's it!
+
+Rust makes for an all-around great language, from prototyping all the way to production.
+
+Gradual improvement is very pleasant in Rust.
+This way, you get the best of both worlds: you can quickly iterate on your design, but you can also make the code more robust over time.
+You can do all of this without sacrificing the advantages of Rust.
+
+It turns out, even "bad" Rust code is pretty decent compared to code I wrote in other languages; it's still safe and fast, covers most cases, and I have an easier time refactoring it later because flaws are more obvious (capital-case types, missing error handling, explicit allocations with `Box` and all).
```

Comment:
Got an article about the joys and ease of refactoring in rust? Would be great to link to it here

Reply:
Not yet, but it would be a good addition.

Reply:
I've added it to my to-do list.